---
title: "ANÁLISIS DE MICROBIOMA"
author: "Gabriela Torres"
date: "2023-08-23"
output:
  html_document:
    toc: true
    number_sections: true
---

# Preparación

## Objeto phyloseq
```{r}
load("~/tesis/01_Data/ps.RData")
```
Cargamos nuestro objeto phyloseq, el cual contiene los ASV, así como una tabla OTU y la asignación taxonómica hasta nivel genéro

## Paquetes a utilizar
```{r}
library(phyloseq)
library(DT)
library(ggplot2)
library(viridis)
library(vegan)
library(microbiome)
library(pheatmap)
library(ggpubr)
library(tidyverse)
library(fantaxtic)
library(DESeq2)
```

# ANALISIS DE MICROBIOTA
## Control de calidad del análisis de 16S
### Prevalencia de los features taxonómicos
```{r}
# data frame que tiene la prevalencia de cada feature  
prevdf = apply(X = otu_table(ps),
               MARGIN = ifelse(taxa_are_rows(ps), yes = 1, no =2),
               FUN = function(x){sum(x > 0)})

# agregamos la taxonomía

prevdf = data.frame(Prevalence = prevdf,
                    TotalAbundance = taxa_sums(ps),
                    tax_table(ps))

plyr::ddply(prevdf, "Genus", function(df1){cbind(mean(df1$Prevalence),sum(df1$Prevalence))}) -> dfprev
datatable(dfprev)
```

```{r}
# Filtramos taxa de acuerdo a un umbral de 1e-5
psd2 <- filter_taxa(ps, function(x) mean(x) > 1e-5, TRUE)

# Filtrar muestras con menos de 500 reads
psd4 <- prune_samples(sample_sums(psd2) > 500, psd2)

psd4
```

```{r}
# Seleccionamos las taxa de interés
prevdf1= subset(prevdf, Order %in% get_taxa_unique(psd4, "Order"))
ggplot(prevdf1, aes(TotalAbundance, Prevalence / nsamples(ps),color=Order)) +

# Agregamos una línea para nuestro umbral
  geom_hline(yintercept = 0.05, alpha = 0.5, linetype = 2) +  geom_point(size = 2, alpha = 0.7) +
  scale_x_log10() +  xlab("Total Abundance") + ylab("Prevalence [Frac. Samples]") +
  facet_wrap(~Phylum) + theme(legend.position="none") + scale_fill_manual(values = viridis(8))
ggsave("~/tesis/03_Output/Prevalencia_Genus.png", dpi=600,width = 21,
  height = 13)
```

```{r}
### GRUPO MDD
p_mdd<- subset_samples(ps, groups == "MDD")
#
prevdf_mdd= apply(X = otu_table(p_mdd),
               MARGIN = ifelse(taxa_are_rows(p_mdd), yes = 1, no =2),
               FUN = function(x){sum(x > 0)})

prevdf_mdd = data.frame(Prevalence = prevdf_mdd,
                    TotalAbundance = taxa_sums(p_mdd),
                    tax_table(p_mdd))

plyr::ddply(prevdf_mdd, "Family", function(df1){cbind(mean(df1$Prevalence),sum(df1$Prevalence))}) -> dfprev_mdd
datatable(dfprev_mdd_new)
#
psd2_mdd <- filter_taxa(p_mdd, function(x) mean(x) > 1e-5, TRUE)

psd4_mdd <- prune_samples(sample_sums(psd2_mdd) > 500, psd2_mdd)
#
prevdf1_mdd = subset(prevdf_mdd, Family %in% get_taxa_unique(psd4_mdd, "Family"))
ggplot(prevdf1_mdd, aes(TotalAbundance, Prevalence / nsamples(p_mdd),color=Genus)) +
  geom_hline(yintercept = 0.01, alpha = 0.5, linetype = 2) +  geom_point(size = 2, alpha = 0.7) +
  scale_x_log10() +  xlab("Total Abundance") + ylab("Prevalence [Frac. Samples]") +
  facet_wrap(~Family) + theme(legend.position="none") + scale_fill_manual(values = viridis(8)) 
ggsave("~/tesis/03_Output/Prevalencia_Family_MDD.png", dpi=600,width = 21,
  height = 13)
```

```{r}
### Grupo CS
p_cs <- subset_samples(ps, groups == "CS")
#
prevdf_cs= apply(X = otu_table(p_cs),
               MARGIN = ifelse(taxa_are_rows(p_cs), yes = 1, no =2),
               FUN = function(x){sum(x > 0)})

prevdf_cs= data.frame(Prevalence = prevdf_cs,
                    TotalAbundance = taxa_sums(p_cs),
                    tax_table(p_cs))

plyr::ddply(prevdf_cs, "Family", function(df1){cbind(mean(df1$Prevalence),sum(df1$Prevalence))}) -> dfprev_cs
datatable(dfprev_cs)
#
psd2_cs <- filter_taxa(p_cs, function(x) mean(x) > 1e-5, TRUE)

psd4_cs<- prune_samples(sample_sums(psd2_cs) > 500, psd2_cs)
#
prevdf1_cs = subset(prevdf_cs, Family %in% get_taxa_unique(psd4_cs, "Family"))
ggplot(prevdf1_cs, aes(TotalAbundance, Prevalence / nsamples(p_cs),color=Genus)) +
  geom_hline(yintercept = 0.01, alpha = 0.5, linetype = 2) +  geom_point(size = 2, alpha = 0.7) +
  scale_x_log10() +  xlab("Total Abundance") + ylab("Prevalence [Frac. Samples]") +
  facet_wrap(~Family) + theme(legend.position="none") + scale_fill_manual(values = viridis(8)) 
ggsave("~/tesis/03_Output/Prevalencia_Family_CS.png", height = 13, width=21,dpi=600)
```

### Curvas de rarefacción

```{r message=FALSE, warning=FALSE}
# Primero cargamos algunos scripts de manera remota
scripts <- c("graphical_methods.R",
             "tree_methods.R",
             "plot_merged_trees.R",
             "specificity_methods.R",
             "ternary_plot.R",
             "richness.R",
             "edgePCA.R",
             "copy_number_correction.R",
             "import_frogs.R",
             "prevalence.R",
             "compute_niche.R")
urls <- paste0("https://raw.githubusercontent.com/mahendra-mariadassou/phyloseq-extended/master/R/", scripts)

for (url in urls) {
  source(url)
}
```

```{r}
#por grupos
p1 <- ggrare(psd4, step = 100, color = "groups", label = "samples", se = TRUE)
p1 <- p1 + facet_wrap(~groups) + scale_color_manual(values = c("#EE7600", "#008B8B")) + scale_fill_manual(values = c("#EE7600", "#008B8B"))
p1
ggsave("~/tesis/03_Output/Rarefactio_Groups.png", dpi=600,width = 21,
  height = 13)
```

## Analisis de diversidad 

### Riqueza (diversidad alfa)

```{r}
p_0week <- subset_samples(psd4, weeks == "0_Weeks") 

a_my_comparisons <- list( c("CS", "MDD"))
symnum.args = list(cutpoints = c(0, 0.0001, 0.001, 0.01, 0.05, 1), symbols = c("****", "***", "**", "*", "ns"))

plot_richness(p_0week, color = "groups", x = "groups", measures = c("Observed", "Chao1", "Shannon","Simpson")) + geom_boxplot(aes(fill = groups, alpha=.7)) + scale_color_manual(values = c("#EE7600", "#008B8B")) + scale_fill_manual(values = c("#EE7600", "#008B8B"))+
  stat_compare_means(method = "wilcox.test", comparisons = a_my_comparisons, label = "p.signif", symnum.args = symnum.args)
ggsave("~/tesis/03_Output/diversidadalfa.png", dpi=600,width = 21,
  height = 13)
```

```{r}
### estimacion de P-value

richness <- estimate_richness(p_0week)
head(richness)

w1 = wilcox.test(richness$Observed ~ sample_data(p_0week)$groups)
w1

w2 = wilcox.test(richness$Chao1 ~ sample_data(p_0week)$groups)
w2

w3 = wilcox.test(richness$Shannon ~ sample_data(p_0week)$groups)
w3

w4 = wilcox.test(richness$Simpson ~ sample_data(p_0week)$groups)
w4
```

```{r}
a_my_comparisons <- list( c("0_Weeks", "08_Weeks"))
symnum.args = list(cutpoints = c(0, 0.0001, 0.001, 0.01, 0.05, 1), symbols = c("****", "***", "**", "*", "ns"))

plot_richness(psd4_mdd, color = "weeks", x = "weeks", measures = c("Observed", "Chao1", "Shannon","Simpson")) + geom_boxplot(aes(fill = weeks, alpha=.7)) + scale_color_manual(values = c("#00868B", "#8B3A62")) + scale_fill_manual(values = c("#00868B", "#8B3A62"))+
  stat_compare_means(method = "wilcox.test", comparisons = a_my_comparisons, label = "p.signif", symnum.args = symnum.args)
ggsave("~/tesis/03_Output/diversidadalfa_mdd.png", height = 13, width=21,dpi=600)
```

```{r}
### estimacion de P-value

richness <- estimate_richness(psd4_mdd)
head(richness)

x1 = wilcox.test(richness$Observed ~ sample_data(psd4_mdd)$weeks)
x1

x2 = wilcox.test(richness$Chao1 ~ sample_data(psd4_mdd)$weeks)
x2

x3 = wilcox.test(richness$Shannon ~ sample_data(psd4_mdd)$weeks)
x3

x4 = wilcox.test(richness$Simpson ~ sample_data(psd4_mdd)$weeks)
x4
```

### Diversidad beta 


